# Capstone Project - Best Restaurant in the Area
### Applied Data Science Capstone

# Introduction

In this project, I will try to find the **best restaurant** in the proximity of UC Davis, San Francisco Campus. This is where the MSBA classes are held, which shares the location with UC Hastings College of the Law. This report will be targeted to prospective/current MSBA students of UC Davis, who might be new to the location and wanting to find good (or palatable) food nearby.

Although I have not been to the vicinity of Civic Centre, yet I believe there should be a lot of restaurant nearby, considering that it is the centre of the city. With so many options, it will be of interests to the students which restaurant is best rated, within the vicinity of the campus.

The objective is to make recommendations to the prospective students’ popular places where they could dine at. I am particularly interested in this project as I will be studying at that campus, assuming everything goes according to plan. I am interested to know what food options are available to me when I grab some food in between classes.

# Data Description

Based on the definition of our problem, we would like to find out if there are any restaurant with the following two criteria:
<li> High ratings </li>
<li> Within proximity to the UC Hastings College of the Law </li> 

As how good a restaurant is varied and largely subjective, with many factors in consideration, such as quality and speed of service, quality of food, cleaniness, consistency etc. Due to the subjectivity, a **rating above 8.0** would be seen as high. 

I would first look at the region that are within **300 metres** to the campus as proximity is an important factor. If no restaurant meets the condition established above, the distance will be increased to **1000 metres**.

When the list of restaurants has been populated, I will take a deeper look at the reviews to delve deeper into why they are highly rated.

I need to extract the following information:  
- The **latitude and longitude** of **UC Hastings College of the Law** will be extracted using geocoding web services offered by **geopy**
- The **restaurants data and location** will be extracted using Foursquare API 

## Methodology

### Methodology Overview

In this project, I will find out which restaurant are good in the vincinity of the campus. I will limit the analysis to area of less than **1000 metres** around UC Hastings College of the Law, as the assumption is that the restaurant should be reachable via foot, without resorting to other means of transport.

In the first part, I will collect the required data location of the campus to set the center of the radius.

As we are travelling on foot, it is more desirable if the restaurant is nearer. Hence, for the second step of my analysis, I would like to find out what are the restaurant that are within **300 metres** from the UC Hastings College of the Law. Moreover, there might not be sufficient time to travel far in-between classes, making grabbing food from a faraway place infeasible. From the query result, I would like to know if there are restaurant options that can return a rating of above **8.0**. If there is an option, I will delve deeper into that option. 

Or else, I will proceed on to query for restaurants within **1000 metre** radius, and that would be the third step. Optimally, I anticipate several restaurants having a rating above **8.0**, as the UC Hastings College of the Law is located at the Civic Centre. Nonetheless, if my hypothesis were to prove false, I will look deeper at what led to its occurrence.  

### Data Sources

Let's import the necessary libraries

In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


### Data Sources - Foursquare
I am interested in venues that contains the word restaurant. I will use Foursquare API to get info on restaurants nearby.

First, lets use my Foursquare Credentials:

In [2]:
CLIENT_ID = 'D5QQI5RXHLN0BBOHU01N4LLXMURCY3XMXVY3NG3UBJG03IHC' # your Foursquare ID
CLIENT_SECRET = 'RTMFBRH2D42JV40A0G2IWVY0VKJMRLRZVTJXDVGFIKO0XOJ2' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 30
print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: D5QQI5RXHLN0BBOHU01N4LLXMURCY3XMXVY3NG3UBJG03IHC
CLIENT_SECRET:RTMFBRH2D42JV40A0G2IWVY0VKJMRLRZVTJXDVGFIKO0XOJ2


Let's first find the latitude & longitude of UC Davis SF Campus, which is sharing its Campus with UC Hastings College of the Law

In [3]:
address = '200 McAllister St, San Francisco, CA 94102, US'

geolocator = Nominatim(user_agent="ucd_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geographical coordinate of UCD are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of UCD are 37.7808388, -122.4156392.


Let's visualise the location of UC Hastings College of the Law on a San Francisco Map:

In [5]:
# create map of Civic Center using latitude and longitude values
map_ucd = folium.Map(location=[latitude, longitude], zoom_start=16)
    
# add a red circle marker to represent the UC Hastings College of the Law
folium.features.CircleMarker(
    [latitude, longitude],
    radius=10,
    color='red',
    popup='UC Hastings College of the Law',
    fill = True,
    fill_color = 'red',
    fill_opacity = 0.6
).add_to(map_ucd)

map_ucd

We can see the UC Hastings College of the Law is indicated by the red icon in the map, it is located in the neighbourhood of the Civic Center, which is a bustling area. I am sure that there will be a lot of good food options nearby. :)

# Analysis

### Analysis - 300 metres

 I want to find some food options nearby in between classes. So, let's define a query to search for restaurant that is within **300 metres** from the UC Hastings College of the Law. 

In [7]:
search_query = 'Restaurant'
radius = 300
print(search_query + ' .... OK!')

Restaurant .... OK!


Define the corresponding URL

In [8]:
url = 'https://api.foursquare.com/v2/venues/search?client_id={}&client_secret={}&ll={},{}&v={}&query={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, latitude, longitude, VERSION, search_query, radius, LIMIT)
url

'https://api.foursquare.com/v2/venues/search?client_id=D5QQI5RXHLN0BBOHU01N4LLXMURCY3XMXVY3NG3UBJG03IHC&client_secret=RTMFBRH2D42JV40A0G2IWVY0VKJMRLRZVTJXDVGFIKO0XOJ2&ll=37.7808388,-122.4156392&v=20180605&query=Restaurant&radius=300&limit=30'

Send the GET Request and examine the results

In [9]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5e92bb1647b43d002b82e2a5'},
 'response': {'venues': [{'id': '4e9f5aec77c8b0898381ce89',
    'name': "Quintero's Restaurant",
    'location': {'address': '393 Eddy St',
     'crossStreet': 'btwn Leavenworth St & Wagner Aly',
     'lat': 37.78326532583694,
     'lng': -122.4136847231462,
     'labeledLatLngs': [{'label': 'display',
       'lat': 37.78326532583694,
       'lng': -122.4136847231462}],
     'distance': 320,
     'postalCode': '94102',
     'cc': 'US',
     'city': 'San Francisco',
     'state': 'CA',
     'country': 'United States',
     'formattedAddress': ['393 Eddy St (btwn Leavenworth St & Wagner Aly)',
      'San Francisco, CA 94102',
      'United States']},
    'categories': [{'id': '4bf58dd8d48988d1c1941735',
      'name': 'Mexican Restaurant',
      'pluralName': 'Mexican Restaurants',
      'shortName': 'Mexican',
      'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/food/mexican_',
       'suffix': '.png'},
      'p

Get relevant part of JSON and transform it into a *pandas* dataframe

In [10]:
# assign relevant part of JSON to venues
venues = results['response']['venues']

# tranform venues into a dataframe
dataframe = pd.json_normalize(venues)
dataframe.head()

id                       name  \
0  4e9f5aec77c8b0898381ce89      Quintero's Restaurant   
1  4f32a5c319836c91c7ec23ef        Restaurant Caliente   
2  4f322af819836c91c7bcd984   Civic Restaurant and Bar   
3  4b787e79f964a5201bd22ee3  Golden Kim Tar Restaurant   
4  4f321de219836c91c7b798d7   Pink Flamingo Restaurant   

                                          categories    referralId  hasPerk  \
0  [{'id': '4bf58dd8d48988d1c1941735', 'name': 'M...  v-1586674424    False   
1  [{'id': '4d4b7105d754a06374d81259', 'name': 'F...  v-1586674424    False   
2  [{'id': '4d4b7105d754a06374d81259', 'name': 'F...  v-1586674424    False   
3  [{'id': '4bf58dd8d48988d145941735', 'name': 'C...  v-1586674424    False   
4  [{'id': '4d4b7105d754a06374d81259', 'name': 'F...  v-1586674424    False   

    location.address              location.crossStreet  location.lat  \
0        393 Eddy St  btwn Leavenworth St & Wagner Aly     37.783265   
1   86 McAllister St                               NaN     37.781052   
2  272 McAllister St                               NaN     37.780578   
3      434 Larkin St    btwn Turk St & Golden Gate Ave     37.781925   
4      448 Larkin St                               NaN     37.781922   

   location.lng                            location.labeledLatLngs  \
0   -122.413685  [{'label': 'display', 'lat': 37.78326532583694...   
1   -122.413371  [{'label': 'display', 'lat': 37.781052, 'lng':...   
2   -122.416420  [{'label': 'display', 'lat': 37.780578, 'lng':...   
3   -122.417351  [{'label': 'display', 'lat': 37.78192531938754...   
4   -122.417266  [{'label': 'display', 'lat': 37.781922, 'lng':...   

   location.distance location.postalCode location.cc  location.city  \
0                320               94102          US  San Francisco   
1                200               94102          US  San Francisco   
2                 74               94102          US  San Francisco   
3                193               94102          US  San Francisco   
4                187               94102          US  San Francisco   

  location.state location.country  \
0             CA    United States   
1             CA    United States   
2             CA    United States   
3             CA    United States   
4             CA    United States   

                           location.formattedAddress location.neighborhood  
0  [393 Eddy St (btwn Leavenworth St & Wagner Aly...                   NaN  
1  [86 McAllister St, San Francisco, CA 94102, Un...                   NaN  
2  [272 McAllister St, San Francisco, CA 94102, U...                   NaN  
3  [434 Larkin St (btwn Turk St & Golden Gate Ave...                   NaN  
4  [448 Larkin St, San Francisco, CA 94102, Unite...                   NaN

Define information of interest and filter dataframe

In [11]:
# keep only columns that include venue name, and anything that is associated with location
filtered_columns = ['name', 'categories'] + [col for col in dataframe.columns if col.startswith('location.')] + ['id']
dataframe_filtered = dataframe.loc[:, filtered_columns]

# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

# filter the category for each row
dataframe_filtered['categories'] = dataframe_filtered.apply(get_category_type, axis=1)

# clean column names by keeping only last term
dataframe_filtered.columns = [column.split('.')[-1] for column in dataframe_filtered.columns]

dataframe_filtered

name          categories            address  \
0        Quintero's Restaurant  Mexican Restaurant        393 Eddy St   
1          Restaurant Caliente                Food   86 McAllister St   
2     Civic Restaurant and Bar                Food  272 McAllister St   
3    Golden Kim Tar Restaurant  Chinese Restaurant      434 Larkin St   
4     Pink Flamingo Restaurant                Food      448 Larkin St   
5         Moon Star Restaurant                Food     1169 Market St   
6             Hai's Restaurant  Chinese Restaurant        227 Hyde St   
7  Cam Hung Chinese Restaurant                Food        294 Turk St   
8          New Star Restaurant  Chinese Restaurant           168 Eddy   
9               Comida Yucatan  Mexican Restaurant        294 Turk St   

                        crossStreet        lat         lng  \
0  btwn Leavenworth St & Wagner Aly  37.783265 -122.413685   
1                               NaN  37.781052 -122.413371   
2                               NaN  37.780578 -122.416420   
3    btwn Turk St & Golden Gate Ave  37.781925 -122.417351   
4                               NaN  37.781922 -122.417266   
5                               NaN  37.779255 -122.414055   
6                  btwn Eddy & Turk  37.783001 -122.415808   
7                               NaN  37.783033 -122.413924   
8                               NaN  37.783776 -122.414171   
9                       Leavenworth  37.782634 -122.415709   

                                      labeledLatLngs  distance postalCode  cc  \
0  [{'label': 'display', 'lat': 37.78326532583694...       320      94102  US   
1  [{'label': 'display', 'lat': 37.781052, 'lng':...       200      94102  US   
2  [{'label': 'display', 'lat': 37.780578, 'lng':...        74      94102  US   
3  [{'label': 'display', 'lat': 37.78192531938754...       193      94102  US   
4  [{'label': 'display', 'lat': 37.781922, 'lng':...       187      94102  US   
5  [{'label': 'display', 'lat': 37.779255, 'lng':...       224      94103  US   
6  [{'label': 'display', 'lat': 37.78300106225665...       241      94102  US   
7  [{'label': 'display', 'lat': 37.783033, 'lng':...       287      94102  US   
8  [{'label': 'display', 'lat': 37.78377635760456...       351      94102  US   
9  [{'label': 'display', 'lat': 37.7826336173429,...       199        NaN  US   

            city state        country  \
0  San Francisco    CA  United States   
1  San Francisco    CA  United States   
2  San Francisco    CA  United States   
3  San Francisco    CA  United States   
4  San Francisco    CA  United States   
5  San Francisco    CA  United States   
6  San Francisco    CA  United States   
7  San Francisco    CA  United States   
8  San Francisco    CA  United States   
9  San Francisco    CA  United States   

                                    formattedAddress neighborhood  \
0  [393 Eddy St (btwn Leavenworth St & Wagner Aly...          NaN   
1  [86 McAllister St, San Francisco, CA 94102, Un...          NaN   
2  [272 McAllister St, San Francisco, CA 94102, U...          NaN   
3  [434 Larkin St (btwn Turk St & Golden Gate Ave...          NaN   
4  [448 Larkin St, San Francisco, CA 94102, Unite...          NaN   
5  [1169 Market St, San Francisco, CA 94103, Unit...          NaN   
6  [227 Hyde St (btwn Eddy & Turk), San Francisco...          NaN   
7  [294 Turk St, San Francisco, CA 94102, United ...          NaN   
8  [168 Eddy, San Francisco, CA 94102, United Sta...          NaN   
9  [294 Turk St (Leavenworth), San Francisco, CA,...   Tenderloin   

                         id  
0  4e9f5aec77c8b0898381ce89  
1  4f32a5c319836c91c7ec23ef  
2  4f322af819836c91c7bcd984  
3  4b787e79f964a5201bd22ee3  
4  4f321de219836c91c7b798d7  
5  4f327a7a19836c91c7dbbcaf  
6  4aef39e1f964a520b6d621e3  
7  4f32500b19836c91c7cb5efb  
8  4deecc9aae609ae4fa3a1991  
9  533e3d0f498ec023c680e8af

In [12]:
dataframe_filtered.name

0          Quintero's Restaurant
1            Restaurant Caliente
2       Civic Restaurant and Bar
3      Golden Kim Tar Restaurant
4       Pink Flamingo Restaurant
5           Moon Star Restaurant
6               Hai's Restaurant
7    Cam Hung Chinese Restaurant
8            New Star Restaurant
9                 Comida Yucatan
Name: name, dtype: object

Let's visualize the restaurants that are nearby

In [13]:
venues_map = folium.Map(location=[latitude, longitude], zoom_start=16) # generate map centred around the Conrad Hotel

# add a red circle marker to represent the UC Hastings College of the Law (UC Davis SF Campus)
folium.features.CircleMarker(
    [latitude, longitude],
    radius=10,
    color='red',
    popup='UC Hastings College of the Law',
    fill = True,
    fill_color = 'red',
    fill_opacity = 0.6
).add_to(venues_map)

# add the restaurants as blue circle markers
for lat, lng, label in zip(dataframe_filtered.lat, dataframe_filtered.lng, dataframe_filtered.categories):
    folium.features.CircleMarker(
        [lat, lng],
        radius=5,
        color='blue',
        popup=label,
        fill = True,
        fill_color='blue',
        fill_opacity=0.6
    ).add_to(venues_map)

# display map
venues_map

I want to frequent the restaurant with the highest ratings

In [14]:
dataframe_filtered.sort_values(by=['distance'])

name          categories            address  \
2     Civic Restaurant and Bar                Food  272 McAllister St   
4     Pink Flamingo Restaurant                Food      448 Larkin St   
3    Golden Kim Tar Restaurant  Chinese Restaurant      434 Larkin St   
9               Comida Yucatan  Mexican Restaurant        294 Turk St   
1          Restaurant Caliente                Food   86 McAllister St   
5         Moon Star Restaurant                Food     1169 Market St   
6             Hai's Restaurant  Chinese Restaurant        227 Hyde St   
7  Cam Hung Chinese Restaurant                Food        294 Turk St   
0        Quintero's Restaurant  Mexican Restaurant        393 Eddy St   
8          New Star Restaurant  Chinese Restaurant           168 Eddy   

                        crossStreet        lat         lng  \
2                               NaN  37.780578 -122.416420   
4                               NaN  37.781922 -122.417266   
3    btwn Turk St & Golden Gate Ave  37.781925 -122.417351   
9                       Leavenworth  37.782634 -122.415709   
1                               NaN  37.781052 -122.413371   
5                               NaN  37.779255 -122.414055   
6                  btwn Eddy & Turk  37.783001 -122.415808   
7                               NaN  37.783033 -122.413924   
0  btwn Leavenworth St & Wagner Aly  37.783265 -122.413685   
8                               NaN  37.783776 -122.414171   

                                      labeledLatLngs  distance postalCode  cc  \
2  [{'label': 'display', 'lat': 37.780578, 'lng':...        74      94102  US   
4  [{'label': 'display', 'lat': 37.781922, 'lng':...       187      94102  US   
3  [{'label': 'display', 'lat': 37.78192531938754...       193      94102  US   
9  [{'label': 'display', 'lat': 37.7826336173429,...       199        NaN  US   
1  [{'label': 'display', 'lat': 37.781052, 'lng':...       200      94102  US   
5  [{'label': 'display', 'lat': 37.779255, 'lng':...       224      94103  US   
6  [{'label': 'display', 'lat': 37.78300106225665...       241      94102  US   
7  [{'label': 'display', 'lat': 37.783033, 'lng':...       287      94102  US   
0  [{'label': 'display', 'lat': 37.78326532583694...       320      94102  US   
8  [{'label': 'display', 'lat': 37.78377635760456...       351      94102  US   

            city state        country  \
2  San Francisco    CA  United States   
4  San Francisco    CA  United States   
3  San Francisco    CA  United States   
9  San Francisco    CA  United States   
1  San Francisco    CA  United States   
5  San Francisco    CA  United States   
6  San Francisco    CA  United States   
7  San Francisco    CA  United States   
0  San Francisco    CA  United States   
8  San Francisco    CA  United States   

                                    formattedAddress neighborhood  \
2  [272 McAllister St, San Francisco, CA 94102, U...          NaN   
4  [448 Larkin St, San Francisco, CA 94102, Unite...          NaN   
3  [434 Larkin St (btwn Turk St & Golden Gate Ave...          NaN   
9  [294 Turk St (Leavenworth), San Francisco, CA,...   Tenderloin   
1  [86 McAllister St, San Francisco, CA 94102, Un...          NaN   
5  [1169 Market St, San Francisco, CA 94103, Unit...          NaN   
6  [227 Hyde St (btwn Eddy & Turk), San Francisco...          NaN   
7  [294 Turk St, San Francisco, CA 94102, United ...          NaN   
0  [393 Eddy St (btwn Leavenworth St & Wagner Aly...          NaN   
8  [168 Eddy, San Francisco, CA 94102, United Sta...          NaN   

                         id  
2  4f322af819836c91c7bcd984  
4  4f321de219836c91c7b798d7  
3  4b787e79f964a5201bd22ee3  
9  533e3d0f498ec023c680e8af  
1  4f32a5c319836c91c7ec23ef  
5  4f327a7a19836c91c7dbbcaf  
6  4aef39e1f964a520b6d621e3  
7  4f32500b19836c91c7cb5efb  
0  4e9f5aec77c8b0898381ce89  
8  4deecc9aae609ae4fa3a1991

Looking good. So now we have all the restaurants in area within few kilometers from Alexanderplatz, and we know which ones are Italian restaurants! We also know which restaurants exactly are in vicinity of every neighborhood candidate center.

This concludes the data gathering phase - we're now ready to use this data for analysis to produce the report on optimal locations for a new Italian restaurant!

In [11]:
# obtain every single list 
id_list = dataframe_filtered['id'].tolist()
id_list

['4e9f5aec77c8b0898381ce89',
 '4f32a5c319836c91c7ec23ef',
 '4f322af819836c91c7bcd984',
 '4b787e79f964a5201bd22ee3',
 '4f321de219836c91c7b798d7',
 '4f327a7a19836c91c7dbbcaf',
 '4aef39e1f964a520b6d621e3',
 '4f32500b19836c91c7cb5efb',
 '4deecc9aae609ae4fa3a1991',
 '533e3d0f498ec023c680e8af']

Let's explore the closest restaurant and Get the Venue's overall rating

In [12]:
venue_id = id_list[0] # ID of Civic Restaurant and Bar
url_food = 'https://api.foursquare.com/v2/venues/{}?client_id={}&client_secret={}&v={}'.format(venue_id, CLIENT_ID, CLIENT_SECRET, VERSION)

result = requests.get(url_food).json()
try:
    print(result['response']['venue']['rating'])
except:
    print('This venue has not been rated yet.')

This venue has not been rated yet.


This restaurant has no ratings... 

Hmm...This does not bode well... how about lets do a for loop to save sometime?  

Let's first generate all urls from the id_list using a for loop:

In [13]:
url_food1 = []
for x in id_list:
    url_food1.append('https://api.foursquare.com/v2/venues/{}?client_id={}&client_secret={}&v={}'.format(x, CLIENT_ID, CLIENT_SECRET, VERSION))
url_food1

['https://api.foursquare.com/v2/venues/4e9f5aec77c8b0898381ce89?client_id=D5QQI5RXHLN0BBOHU01N4LLXMURCY3XMXVY3NG3UBJG03IHC&client_secret=RTMFBRH2D42JV40A0G2IWVY0VKJMRLRZVTJXDVGFIKO0XOJ2&v=20180605',
 'https://api.foursquare.com/v2/venues/4f32a5c319836c91c7ec23ef?client_id=D5QQI5RXHLN0BBOHU01N4LLXMURCY3XMXVY3NG3UBJG03IHC&client_secret=RTMFBRH2D42JV40A0G2IWVY0VKJMRLRZVTJXDVGFIKO0XOJ2&v=20180605',
 'https://api.foursquare.com/v2/venues/4f322af819836c91c7bcd984?client_id=D5QQI5RXHLN0BBOHU01N4LLXMURCY3XMXVY3NG3UBJG03IHC&client_secret=RTMFBRH2D42JV40A0G2IWVY0VKJMRLRZVTJXDVGFIKO0XOJ2&v=20180605',
 'https://api.foursquare.com/v2/venues/4b787e79f964a5201bd22ee3?client_id=D5QQI5RXHLN0BBOHU01N4LLXMURCY3XMXVY3NG3UBJG03IHC&client_secret=RTMFBRH2D42JV40A0G2IWVY0VKJMRLRZVTJXDVGFIKO0XOJ2&v=20180605',
 'https://api.foursquare.com/v2/venues/4f321de219836c91c7b798d7?client_id=D5QQI5RXHLN0BBOHU01N4LLXMURCY3XMXVY3NG3UBJG03IHC&client_secret=RTMFBRH2D42JV40A0G2IWVY0VKJMRLRZVTJXDVGFIKO0XOJ2&v=20180605',
 'htt

Then find out the ratings of all restaurants using a for loop

In [14]:
for x in url_food1:
    result = requests.get(x).json()
    try:
        print(result['response']['venue']['rating'])
    except:
        print('This venue has not been rated yet.')

This venue has not been rated yet.
This venue has not been rated yet.
This venue has not been rated yet.
7.4
This venue has not been rated yet.
This venue has not been rated yet.
This venue has not been rated yet.
This venue has not been rated yet.
This venue has not been rated yet.
This venue has not been rated yet.


Hmm... Only 1 venue has been rated.

Let's take a quick look at the only restaurant that has a rating, which is Golden Kim Tar Restaurant, before expanding our search to 1000 metres around the vicinity.

Get the number of tips of the restaurant

In [15]:
url_food = 'https://api.foursquare.com/v2/venues/{}?client_id={}&client_secret={}&v={}'.format('4b787e79f964a5201bd22ee3', CLIENT_ID, CLIENT_SECRET, VERSION)
result = requests.get(url_food).json()
result['response']['venue']['tips']['count']

19

19 Reviews, this is not a lot, lets see what exactly the reviewer says, it could be a hidden gem in disguise.

##### Get the venue's tips

Create URL and send GET request

In [18]:
limit = 20 # set limit to be greater than or equal to the total number of tips
url_food = 'https://api.foursquare.com/v2/venues/{}/tips?client_id={}&client_secret={}&v={}&limit={}'.format('4b787e79f964a5201bd22ee3', CLIENT_ID, CLIENT_SECRET, VERSION, limit)

results = requests.get(url_food).json()
results

{'meta': {'code': 200, 'requestId': '5e8ffe991835dd001b8a0f26'},
 'response': {'tips': {'count': 19,
   'items': [{'id': '51fd7d91498ea69707a837d0',
     'createdAt': 1375567249,
     'text': 'Coworkers brought me here. Great value, big portions, VERY ,fast service. Ordered the broccoli beef, my coworkers swear by the salt & pepper chicken wings..... Will try that next.',
     'type': 'user',
     'canonicalUrl': 'https://foursquare.com/item/51fd7d91498ea69707a837d0',
     'lang': 'en',
     'likes': {'count': 1,
      'groups': [{'type': 'others',
        'count': 1,
        'items': [{'id': '9129717',
          'firstName': 'Judy',
          'photo': {'prefix': 'https://fastly.4sqi.net/img/user/',
           'suffix': '/FTPBKRP01JECAKXI.jpg'}}]}],
      'summary': '1 like'},
     'logView': True,
     'agreeCount': 1,
     'disagreeCount': 0,
     'todo': {'count': 0},
     'user': {'id': '2558780',
      'firstName': 'Kay',
      'lastName': 'V',
      'photo': {'prefix': 'https://f

###### Get tips and list of associated features

In [19]:
tips = results['response']['tips']['items']

tip = results['response']['tips']['items'][0]
tip.keys()

dict_keys(['id', 'createdAt', 'text', 'type', 'canonicalUrl', 'lang', 'likes', 'logView', 'agreeCount', 'disagreeCount', 'todo', 'user'])

Format column width and display all tips

In [20]:
pd.set_option('display.max_colwidth', None)

tips_df = pd.json_normalize(tips) # json normalize tips
tips_df.head()

id   createdAt  \
0  51fd7d91498ea69707a837d0  1375567249   

                                                                                                                                                                                  text  \
0  Coworkers brought me here. Great value, big portions, VERY ,fast service. Ordered the broccoli beef, my coworkers swear by the salt & pepper chicken wings..... Will try that next.   

   type                                          canonicalUrl lang  logView  \
0  user  https://foursquare.com/item/51fd7d91498ea69707a837d0   en     True   

   agreeCount  disagreeCount  likes.count  \
0           1              0            1   

                                                                                                                                                                       likes.groups  \
0  [{'type': 'others', 'count': 1, 'items': [{'id': '9129717', 'firstName': 'Judy', 'photo': {'prefix': 'https://fastly.4sqi.net/img/user/', 'suffix': '/FTPBKRP01JECAKXI.jpg'}}]}]   

  likes.summary  todo.count  user.id user.firstName user.lastName  \
0        1 like           0  2558780            Kay             V   

                   user.photo.prefix  \
0  https://fastly.4sqi.net/img/user/   

                                                                        user.photo.suffix  
0  /2558780_leEzNiiT_SQ87I6cH_mE0PofwCNqvvBcJGvUYHUQw42WKCO_SA_Z-kVsDmmD0jWT-VWYFAh9j.jpg

Let's clean it up

In [21]:
# columns to keep
# filtered_columns = ['text', 'agreeCount', 'disagreeCount', 'id', 'user.firstName', 'user.lastName','user.id']
tips_filtered = tips_df[['text', 'agreeCount', 'disagreeCount', 'id', 'user.firstName', 'user.lastName', 'user.id']]

# display tips
tips_filtered

text  \
0  Coworkers brought me here. Great value, big portions, VERY ,fast service. Ordered the broccoli beef, my coworkers swear by the salt & pepper chicken wings..... Will try that next.   

   agreeCount  disagreeCount                        id user.firstName  \
0           1              0  51fd7d91498ea69707a837d0            Kay   

  user.lastName  user.id  
0             V  2558780

Although I can only see one comment (due to the account restriction), but it seems great. The students are definitely interested in great value options and fast service. It sounds like a great place for students, I will try out that place in the future, especially when I am in between classes. However, since it did not meet the criteria, let us expand our search zone and seek better options. 

### Analysis - Expanding to 1000 metres radius

Now I will explore the options available when extending the range to 1000 metres radius 

In [15]:
search_query = 'Restaurant'
radius = 1000
print(search_query + ' .... OK!')

Restaurant .... OK!


Define the url for query and query

In [16]:
url = 'https://api.foursquare.com/v2/venues/search?client_id={}&client_secret={}&ll={},{}&v={}&query={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, latitude, longitude, VERSION, search_query, radius, LIMIT)
results = requests.get(url).json()

Transform the results into a dataframe

In [17]:
# assign relevant part of JSON to venues
venues = results['response']['venues']

# tranform venues into a dataframe
dataframe = pd.json_normalize(venues)
dataframe.head()

id                                  name  \
0  43f4dc37f964a5209f2f1fe3               Turtle Tower Restaurant   
1  41799f00f964a520c11d1fe3                     Tú Lan Restaurant   
2  435c2400f964a5200e291fe3           Basil Thai Restaurant & Bar   
3  42717900f964a5206a211fe3  The Chieftain Irish Pub & Restaurant   
4  433c8000f964a52043281fe3     Ryoko's Japanese Restaurant & Bar   

                                          categories    referralId  hasPerk  \
0  [{'id': '4bf58dd8d48988d14a941735', 'name': 'V...  v-1586674529    False   
1  [{'id': '4bf58dd8d48988d14a941735', 'name': 'V...  v-1586674529    False   
2  [{'id': '4bf58dd8d48988d149941735', 'name': 'T...  v-1586674529    False   
3  [{'id': '52e81612bcbc57f1066b7a06', 'name': 'I...  v-1586674529    False   
4  [{'id': '4bf58dd8d48988d1d2941735', 'name': 'S...  v-1586674529    False   

  location.address           location.crossStreet  location.lat  location.lng  \
0    645 Larkin St              btwn Eddy & Ellis     37.783989   -122.417764   
1         8 6th St  btwn Market St & Stevenson St     37.781987   -122.409977   
2   1175 Folsom St                      at Eighth     37.775388   -122.409253   
3       198 5th St                   at Howard St     37.781502   -122.404972   
4    619 Taylor St                    at Cosmo Pl     37.788183   -122.411882   

                             location.labeledLatLngs  location.distance  \
0  [{'label': 'display', 'lat': 37.78398947390574...                397   
1  [{'label': 'display', 'lat': 37.78198675959605...                514   
2  [{'label': 'display', 'lat': 37.775388, 'lng':...                826   
3  [{'label': 'display', 'lat': 37.78150216218952...                941   
4  [{'label': 'display', 'lat': 37.788183, 'lng':...                881   

  location.postalCode location.cc  location.city location.state  \
0               94109          US  San Francisco             CA   
1               94103          US  San Francisco             CA   
2               94103          US  San Francisco             CA   
3               94103          US  San Francisco             CA   
4               94102          US  San Francisco             CA   

  location.country                          location.formattedAddress  \
0    United States  [645 Larkin St (btwn Eddy & Ellis), San Franci...   
1    United States  [8 6th St (btwn Market St & Stevenson St), San...   
2    United States  [1175 Folsom St (at Eighth), San Francisco, CA...   
3    United States  [198 5th St (at Howard St), San Francisco, CA ...   
4    United States  [619 Taylor St (at Cosmo Pl), San Francisco, C...   

  venuePage.id delivery.id                                       delivery.url  \
0     94430754         NaN                                                NaN   
1          NaN     2054098  https://www.grubhub.com/restaurant/t-lan-8-6th...   
2          NaN         NaN                                                NaN   
3     96608118         NaN                                                NaN   
4    105178475         NaN                                                NaN   

  delivery.provider.name             delivery.provider.icon.prefix  \
0                    NaN                                       NaN   
1                grubhub  https://fastly.4sqi.net/img/general/cap/   
2                    NaN                                       NaN   
3                    NaN                                       NaN   
4                    NaN                                       NaN   

  delivery.provider.icon.sizes              delivery.provider.icon.name  \
0                          NaN                                      NaN   
1                     [40, 50]  /delivery_provider_grubhub_20180129.png   
2                          NaN                                      NaN   
3                          NaN                                      NaN   
4                          NaN                                      NaN 

Clean the dataframe

In [18]:
# keep only columns that include venue name, and anything that is associated with location
filtered_columns = ['name', 'categories'] + [col for col in dataframe.columns if col.startswith('location.')] + ['id']
dataframe_filtered = dataframe.loc[:, filtered_columns]

# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

# filter the category for each row
dataframe_filtered['categories'] = dataframe_filtered.apply(get_category_type, axis=1)

# clean column names by keeping only last term
dataframe_filtered.columns = [column.split('.')[-1] for column in dataframe_filtered.columns]

dataframe_filtered

name  \
0                           Turtle Tower Restaurant   
1                                 Tú Lan Restaurant   
2                       Basil Thai Restaurant & Bar   
3              The Chieftain Irish Pub & Restaurant   
4                 Ryoko's Japanese Restaurant & Bar   
5                          Pakwan Indian Restaurant   
6                              7 Mission Restaurant   
7          Restaurant & Hotel Upholstery Specialist   
8                                Chutney Restaurant   
9                                   Restaurant Anzu   
10                   Don Ramon's Mexican Restaurant   
11                              Yemeni's Restaurant   
12                            Quintero's Restaurant   
13                                  Fino Restaurant   
14                           Cafe Splash Restaurant   
15                           MKT Restaurant and Bar   
16         The Restaurant Collection Under The Dome   
17  Eden's Mediterranean Turkish & Greek Restaurant   
18                       Java Indonesian Restaurant   
19                                Darbar Restaurant   
20                              Restaurant Caliente   
21                         Civic Restaurant and Bar   
22                  Thai Idea Vegetarian Restaurant   
23                               B Restaurant & Bar   
24                        Golden Kim Tar Restaurant   
25                            A La Turca Restaurant   
26                       Cityscape Bar & Restaurant   
27        The Restaurant at the End of the Universe   
28                                 Restaurant Depot   
29                         Teo's Restaurant and Bar   

                       categories            address  \
0           Vietnamese Restaurant      645 Larkin St   
1           Vietnamese Restaurant           8 6th St   
2                 Thai Restaurant     1175 Folsom St   
3                       Irish Pub         198 5th St   
4                Sushi Restaurant      619 Taylor St   
5               Indian Restaurant    501 Ofarrell St   
6              Chinese Restaurant         150 7th St   
7          Furniture / Home Store      120 Willow St   
8               Indian Restaurant       511 Jones St   
9             Japanese Restaurant       222 Mason St   
10             Mexican Restaurant        225 11th St   
11      Middle Eastern Restaurant     1098 Sutter St   
12             Mexican Restaurant        393 Eddy St   
13             Italian Restaurant        624 Post St   
14          Vietnamese Restaurant       750 Ellis St   
15                     Steakhouse      757 Market St   
16        New American Restaurant                NaN   
17       Mediterranean Restaurant       552 Jones St   
18          Indonesian Restaurant    3015 Geary Blvd   
19              Indian Restaurant       1412 Polk St   
20                           Food   86 McAllister St   
21                           Food  272 McAllister St   
22  Vegetarian / Vegan Restaurant        710 Polk St   
23                         Lounge      720 Howard St   
24             Chinese Restaurant      434 Larkin St   
25             Turkish Restaurant       869 Geary St   
26        New American Restaurant    333 Ofarrell St   
27                     Strip Club    935 Ofarrell St   
28                           None                NaN   
29             Chinese Restaurant         112 7th St   

                         crossStreet        lat         lng  \
0                  btwn Eddy & Ellis  37.783989 -122.417764   
1      btwn Market St & Stevenson St  37.781987 -122.409977   
2                          at Eighth  37.775388 -122.409253   
3                       at Howard St  37.781502 -122.404972   
4                        at Cosmo Pl  37.788183 -122.411882   
5                        at Jones St  37.785752 -122.412945   
6              btwn Mission & Howard  37.778571 -122.410195   
7                                NaN  37.783722 -122.419803   
8                    at O'Farrell St 

What are the names of the restaurants

In [19]:
dataframe_filtered.name

0                             Turtle Tower Restaurant
1                                   Tú Lan Restaurant
2                         Basil Thai Restaurant & Bar
3                The Chieftain Irish Pub & Restaurant
4                   Ryoko's Japanese Restaurant & Bar
5                            Pakwan Indian Restaurant
6                                7 Mission Restaurant
7            Restaurant & Hotel Upholstery Specialist
8                                  Chutney Restaurant
9                                     Restaurant Anzu
10                     Don Ramon's Mexican Restaurant
11                                Yemeni's Restaurant
12                              Quintero's Restaurant
13                                    Fino Restaurant
14                             Cafe Splash Restaurant
15                             MKT Restaurant and Bar
16           The Restaurant Collection Under The Dome
17    Eden's Mediterranean Turkish & Greek Restaurant
18                         J

Visualise the restaurants on a map

In [21]:
venues_map = folium.Map(location=[latitude, longitude], zoom_start=14) # generate map centred around the Conrad Hotel

# add a red circle marker to represent the UC Hastings College of the Law (UC Davis SF Campus)
folium.features.CircleMarker(
    [latitude, longitude],
    radius=10,
    color='red',
    popup='UC Hastings College of the Law',
    fill = True,
    fill_color = 'red',
    fill_opacity = 0.6
).add_to(venues_map)

# add the restaurants as blue circle markers
for lat, lng, label in zip(dataframe_filtered.lat, dataframe_filtered.lng, dataframe_filtered.categories):
    folium.features.CircleMarker(
        [lat, lng],
        radius=5,
        color='blue',
        popup=label,
        fill = True,
        fill_color='blue',
        fill_opacity=0.6
    ).add_to(venues_map)

# display map
venues_map

Save the IDs to a list

In [29]:
# obtain every single list 
id_list_2 = dataframe_filtered['id'].tolist()
id_list_2

['43f4dc37f964a5209f2f1fe3',
 '41799f00f964a520c11d1fe3',
 '435c2400f964a5200e291fe3',
 '42717900f964a5206a211fe3',
 '433c8000f964a52043281fe3',
 '4ad4e1f7f964a52069fc20e3',
 '4a9323a3f964a520ea1e20e3',
 '5a1ffa301987ec7848143581',
 '49efdfa9f964a52015691fe3',
 '4ac44acdf964a520949e20e3',
 '49f0fec0f964a5207b691fe3',
 '4e9f5aec77c8b0898381ce89',
 '4ba3db3af964a520166738e3',
 '4a19721ef964a520197a1fe3',
 '4b1aba3ff964a52070f023e3',
 '5215838211d299f75876792e',
 '4b72155ff964a520cb6d2de3',
 '49fb9750f964a5205f6e1fe3',
 '45209ad3f964a520b83a1fe3',
 '3fd66200f964a520f6ec1ee3',
 '4b62377ff964a520c53c2ae3',
 '4f32a5c319836c91c7ec23ef',
 '4f322af819836c91c7bcd984',
 '4e7d7aebf5b9644b441e28f3',
 '4b787e79f964a5201bd22ee3',
 '4a6cc33df964a52091d11fe3',
 '4d253f45bcc1224bfbae8f1b',
 '4ff7b450d63e98a680674a36',
 '4838a10af964a520f14f1fe3',
 '57392643498e45f1e3ac4f62']

Use a for loop to generate all urls for query

In [32]:
url_food_2 = []
for x in id_list_2:
    url_food_2.append('https://api.foursquare.com/v2/venues/{}?client_id={}&client_secret={}&v={}'.format(x, CLIENT_ID, CLIENT_SECRET, VERSION))
url_food_2

['https://api.foursquare.com/v2/venues/43f4dc37f964a5209f2f1fe3?client_id=D5QQI5RXHLN0BBOHU01N4LLXMURCY3XMXVY3NG3UBJG03IHC&client_secret=RTMFBRH2D42JV40A0G2IWVY0VKJMRLRZVTJXDVGFIKO0XOJ2&v=20180605',
 'https://api.foursquare.com/v2/venues/41799f00f964a520c11d1fe3?client_id=D5QQI5RXHLN0BBOHU01N4LLXMURCY3XMXVY3NG3UBJG03IHC&client_secret=RTMFBRH2D42JV40A0G2IWVY0VKJMRLRZVTJXDVGFIKO0XOJ2&v=20180605',
 'https://api.foursquare.com/v2/venues/435c2400f964a5200e291fe3?client_id=D5QQI5RXHLN0BBOHU01N4LLXMURCY3XMXVY3NG3UBJG03IHC&client_secret=RTMFBRH2D42JV40A0G2IWVY0VKJMRLRZVTJXDVGFIKO0XOJ2&v=20180605',
 'https://api.foursquare.com/v2/venues/42717900f964a5206a211fe3?client_id=D5QQI5RXHLN0BBOHU01N4LLXMURCY3XMXVY3NG3UBJG03IHC&client_secret=RTMFBRH2D42JV40A0G2IWVY0VKJMRLRZVTJXDVGFIKO0XOJ2&v=20180605',
 'https://api.foursquare.com/v2/venues/433c8000f964a52043281fe3?client_id=D5QQI5RXHLN0BBOHU01N4LLXMURCY3XMXVY3NG3UBJG03IHC&client_secret=RTMFBRH2D42JV40A0G2IWVY0VKJMRLRZVTJXDVGFIKO0XOJ2&v=20180605',
 'htt

Use a for loop to obtain ratings for every restaurant

In [33]:
for x in url_food_2:
    result = requests.get(x).json()
    try:
        print(result['response']['venue']['rating'])
    except:
        print('This venue has not been rated yet.')

8.0
8.5
8.3
7.1
8.9
6.3
7.3
This venue has not been rated yet.
7.6
6.8
7.4
This venue has not been rated yet.
7.3
7.6
This venue has not been rated yet.
This venue has not been rated yet.
7.8
This venue has not been rated yet.
6.4
This venue has not been rated yet.
7.3
This venue has not been rated yet.
This venue has not been rated yet.
7.4
7.4
7.9
This venue has not been rated yet.
This venue has not been rated yet.
This venue has not been rated yet.
This venue has not been rated yet.


This definitely look better than narrowing the search to 1000 metres

The four restaurants that has a rating of above or equal to 8.0 are Turtle Tower Restaurant, Tú Lan Restaurant, Basil Thai Restaurant & Bar and Ryoko's Japanese Restaurant & Bar.

Lets take a look at each of the restaurant respectively:

#### Turtle Tower Restaurant

##### Get the number of tips for the Turtle Tower Restaurant

In [4]:
url_food = 'https://api.foursquare.com/v2/venues/{}?client_id={}&client_secret={}&v={}'.format('43f4dc37f964a5209f2f1fe3', CLIENT_ID, CLIENT_SECRET, VERSION)
result = requests.get(url_food).json()
result['response']['venue']['tips']['count']

117

117 Reviews, this is a good amount, quite a popular restaurant it seems.

##### Get the tips for the Turtle Tower Restaurant

In [38]:
limit = 1 # set limit to be greater than or equal to the total number of tips
url_food0 = 'https://api.foursquare.com/v2/venues/{}/tips?client_id={}&client_secret={}&v={}&limit={}'.format('43f4dc37f964a5209f2f1fe3', CLIENT_ID, CLIENT_SECRET, VERSION, limit)

results0 = requests.get(url_food0).json()

In [39]:
# Get tips and list of associated features
tips0 = results0['response']['tips']['items']

In [40]:
# Format column width and display all tips
pd.set_option('display.max_colwidth', None)

tip0 = pd.json_normalize(tips0) # json normalize tips
tip0.head()

id   createdAt  \
0  545406bb498e43d4ed60d015  1414792891   

                                                                                                                                                                                     text  \
0  Order the Chicken Pho, it's the best in the Bay area with a broth as delicious as it gets. If your up for a challenge add Pho Ga Long to your broth mixing chicken giblets to the mix.   

   type                                                                 url  \
0  user  http://www.thrillist.com/eat/san-francisco/best-pho-in-sf-bay-area   

                                           canonicalUrl  \
0  https://foursquare.com/item/545406bb498e43d4ed60d015   

                                                                                                photourl  \
0  https://fastly.4sqi.net/img/general/original/86922724_APX4WL5LyIQGfqheEF7ob6y3DqQzzJl9adE_hSn6e1U.jpg   

  lang  logView  agreeCount  disagreeCount                  photo.id  \
0   en     True           3              0  545406bc498e32f0806d2ef8   

   photo.createdAt photo.source.name        photo.source.url  \
0       1414792892    Foursquare Web  https://foursquare.com   

                           photo.prefix  \
0  https://fastly.4sqi.net/img/general/   

                                                photo.suffix  photo.width  \
0  /86922724_APX4WL5LyIQGfqheEF7ob6y3DqQzzJl9adE_hSn6e1U.jpg          960   

   photo.height photo.visibility  likes.count likes.groups  todo.count  \
0           720           public            0           []           0   

    user.id user.firstName                  user.photo.prefix  \
0  86922724    Staff Picks  https://fastly.4sqi.net/img/user/   

                user.photo.suffix user.type  
0  /86922724-IJWQLN5RUL3IAIS5.png      page

Lets clean it up

In [41]:
# columns to keep
# filtered_columns = ['text', 'agreeCount', 'disagreeCount', 'id', 'user.firstName', 'user.lastName','user.id']
tips_filtered0 = tip0[['text', 'agreeCount', 'disagreeCount', 'id', 'user.firstName', 'user.id']]

# display tips
tips_filtered0

text  \
0  Order the Chicken Pho, it's the best in the Bay area with a broth as delicious as it gets. If your up for a challenge add Pho Ga Long to your broth mixing chicken giblets to the mix.   

   agreeCount  disagreeCount                        id user.firstName  \
0           3              0  545406bb498e43d4ed60d015    Staff Picks   

    user.id  
0  86922724

Hmm... The review is very positive. If I want to have Chicken Pho, its definitely the place to go. However, it seems that this comment is stickied by the Staff of the restaurant, so it does not surprise me that it is a good review. Regardless, there are 3 agree counts to this, so yeah, it should be a good choice. Lets look at the second option:

#### Tú Lan Restaurant

##### Get the number of tips for the Tú Lan Restaurant

In [42]:
url_food = 'https://api.foursquare.com/v2/venues/{}?client_id={}&client_secret={}&v={}'.format('41799f00f964a520c11d1fe3', CLIENT_ID, CLIENT_SECRET, VERSION)
result = requests.get(url_food).json()
result['response']['venue']['tips']['count']

144

117 Reviews, even more reviews, quite a popular restaurant it seems.

###### Get the tips for the Tú Lan Restaurant

In [43]:
limit = 1 # set limit to be greater than or equal to the total number of tips
url_food1 = 'https://api.foursquare.com/v2/venues/{}/tips?client_id={}&client_secret={}&v={}&limit={}'.format('41799f00f964a520c11d1fe3', CLIENT_ID, CLIENT_SECRET, VERSION, limit)

results1 = requests.get(url_food1).json()

In [44]:
# Get tips and list of associated features
tips1 = results1['response']['tips']['items']

In [45]:
# Format column width and display all tips
pd.set_option('display.max_colwidth', None)

tip1 = pd.json_normalize(tips1) # json normalize tips
tip1.head()

id   createdAt  \
0  56466ebf498e5219f1fc7a9f  1447456447   

                                                                                                                                                              text  \
0  AMAZING VIETNAMESE. I had the pho and the shrimp salad. Super duper tasty. They do take out and you'll have to queue for seats. Always room for one at the bar.   

   type                                          canonicalUrl  \
0  user  https://foursquare.com/item/56466ebf498e5219f1fc7a9f   

                                                                                                 photourl  \
0  https://fastly.4sqi.net/img/general/original/111578617_FkYaz3g16jkPTro7FoXDaKqgPGGYa390jzgglO6wXeQ.jpg   

  lang  logView  agreeCount  disagreeCount    lastVoteText  \
0   en     True           6              0  Upvoted Feb 29   

   lastUpvoteTimestamp authorInteractionType                  photo.id  \
0           1582947944                 liked  56466ec1498ef794fb206433   

   photo.createdAt   photo.source.name  \
0       1447456449  Foursquare for iOS   

                           photo.source.url  \
0  https://foursquare.com/download/#/iphone   

                           photo.prefix  \
0  https://fastly.4sqi.net/img/general/   

                                                 photo.suffix  photo.width  \
0  /111578617_FkYaz3g16jkPTro7FoXDaKqgPGGYa390jzgglO6wXeQ.jpg         1440   

   photo.height photo.visibility  likes.count likes.groups  todo.count  \
0          1440           public            0           []           0   

     user.id user.firstName user.lastName                  user.photo.prefix  \
0  111578617        Harriet             R  https://fastly.4sqi.net/img/user/   

                 user.photo.suffix  
0  /111578617-I10BEGUJR0IVGACT.jpg

In [46]:
# columns to keep
# filtered_columns = ['text', 'agreeCount', 'disagreeCount', 'id', 'user.firstName', 'user.lastName','user.id']
tips_filtered1 = tip1[['text', 'agreeCount', 'disagreeCount', 'id', 'user.firstName', 'user.lastName', 'user.id']]

# display tips
tips_filtered1

text  \
0  AMAZING VIETNAMESE. I had the pho and the shrimp salad. Super duper tasty. They do take out and you'll have to queue for seats. Always room for one at the bar.   

   agreeCount  disagreeCount                        id user.firstName  \
0           6              0  56466ebf498e5219f1fc7a9f        Harriet   

  user.lastName    user.id  
0             R  111578617

Hahaha, another Vietnamese restaurant with great reviews. I will definitely try out the shrimp salad. This is definitely a recommendation. Moving on the third one: 

#### Basil Thai Restaurant & Bar

##### Get the number of tips for the Basil Thai Restaurant & Bar

In [47]:
url_food = 'https://api.foursquare.com/v2/venues/{}?client_id={}&client_secret={}&v={}'.format('435c2400f964a5200e291fe3', CLIENT_ID, CLIENT_SECRET, VERSION)
result = requests.get(url_food).json()
result['response']['venue']['tips']['count']

100

100 Reviews, this is a good amount, quite a popular restaurant it seems.

##### Get the tips for the Basil Thai Restaurant & Bar

In [48]:
limit = 1 # set limit to be greater than or equal to the total number of tips
url_food2 = 'https://api.foursquare.com/v2/venues/{}/tips?client_id={}&client_secret={}&v={}&limit={}'.format('435c2400f964a5200e291fe3', CLIENT_ID, CLIENT_SECRET, VERSION, limit)

results2 = requests.get(url_food2).json()

In [49]:
# Get tips and list of associated features
tips2 = results2['response']['tips']['items']


In [50]:
# Format column width and display all tips
pd.set_option('display.max_colwidth', None)

tip2 = pd.json_normalize(tips2) # json normalize tips
tip2.head()

id   createdAt  \
0  56eb15b4498e7f8447ee7b78  1458247092   

                                                                                                                                                                                             text  \
0  You can almost always walk right in with no wait. My favorite items are the eggplant salad with shrimp and pork, and the peppercorn duck breast. Everything I've tasted has been good, though.   

   type                                          canonicalUrl lang  logView  \
0  user  https://foursquare.com/item/56eb15b4498e7f8447ee7b78   en     True   

   agreeCount  disagreeCount authorInteractionType  likes.count likes.groups  \
0           3              0                 liked            0           []   

   todo.count    user.id user.firstName user.lastName  \
0           0  168464378           Jane             R   

                   user.photo.prefix                user.photo.suffix  
0  https://fastly.4sqi.net/img/user/  /168464378-QTIOFMBRX3YNZ32L.jpg

In [51]:
# columns to keep
# filtered_columns = ['text', 'agreeCount', 'disagreeCount', 'id', 'user.firstName', 'user.lastName','user.id']
tips_filtered2 = tip2[['text', 'agreeCount', 'disagreeCount', 'id', 'user.firstName', 'user.lastName', 'user.id']]

# display tips
tips_filtered2

text  \
0  You can almost always walk right in with no wait. My favorite items are the eggplant salad with shrimp and pork, and the peppercorn duck breast. Everything I've tasted has been good, though.   

   agreeCount  disagreeCount                        id user.firstName  \
0           3              0  56eb15b4498e7f8447ee7b78           Jane   

  user.lastName    user.id  
0             R  168464378

This is a Thai restaurant with seemingly interesting fusion food. Interesting option it is. Moving on to the last option and the most highly rated one.

#### Ryoko's Japanese Restaurant & Bar

##### Get the number of tips for the Ryoko's Japanese Restaurant & Bar

In [33]:
url_food = 'https://api.foursquare.com/v2/venues/{}?client_id={}&client_secret={}&v={}'.format('433c8000f964a52043281fe3', CLIENT_ID, CLIENT_SECRET, VERSION)
result = requests.get(url_food).json()
result['response']['venue']['tips']['count']

439

Wow, so many reviews.

##### Get the tips for the Ryoko's Japanese Restaurant & Bar

In [34]:
limit = 1 # set limit to be greater than or equal to the total number of tips
url_food3 = 'https://api.foursquare.com/v2/venues/{}/tips?client_id={}&client_secret={}&v={}&limit={}'.format('433c8000f964a52043281fe3', CLIENT_ID, CLIENT_SECRET, VERSION, limit)

results3 = requests.get(url_food3).json()

In [35]:
# Get tips and list of associated features
tips3 = results3['response']['tips']['items']

In [36]:
# Format column width and display all tips
pd.set_option('display.max_colwidth', None)

tip3 = pd.json_normalize(tips3) # json normalize tips
tip3.head()

id   createdAt  \
0  5523f574498e1cb7c96aaebc  1428419956   

                                                                                                                                                                                                       text  \
0  When late night munchies strike, there's no better place to be. The fresh fish in the sushi is outstanding and they also have a large number of imports from Tsukiji. Read more on WinstonWanders below!   

   type                                                        url  \
0  user  http://winstonwanders.com/2015/04/06/ryokos-union-square/   

                                           canonicalUrl  \
0  https://foursquare.com/item/5523f574498e1cb7c96aaebc   

                                                                                               photourl  \
0  https://fastly.4sqi.net/img/general/original/4083549_5FUQ0uZ9vQTt12Vq6gunB2k2km0MNLLzZ_AcxrVAPZg.jpg   

  lang  logView  agreeCount  disagreeCount    lastVoteText  \
0   en     True          12              0  Upvoted Feb 22   

   lastUpvoteTimestamp authorInteractionType                  photo.id  \
0           1582338615                 liked  5523f575498ef583b0e232ff   

   photo.createdAt photo.source.name        photo.source.url  \
0       1428419957    Foursquare Web  https://foursquare.com   

                           photo.prefix  \
0  https://fastly.4sqi.net/img/general/   

                                               photo.suffix  photo.width  \
0  /4083549_5FUQ0uZ9vQTt12Vq6gunB2k2km0MNLLzZ_AcxrVAPZg.jpg         1200   

   photo.height photo.visibility  likes.count likes.groups  todo.count  \
0          1200           public            0           []           1   

   user.id user.firstName user.lastName                  user.photo.prefix  \
0  4083549           Mike             W  https://fastly.4sqi.net/img/user/   

               user.photo.suffix  
0  /4083549-14ZMSGT0BA3WO1OP.jpg

In [37]:
# columns to keep
# filtered_columns = ['text', 'agreeCount', 'disagreeCount', 'id', 'user.firstName', 'user.lastName','user.id']
tips_filtered3 = tip3[['text', 'agreeCount', 'disagreeCount', 'id', 'user.firstName', 'user.lastName', 'user.id']]

# display tips
tips_filtered3

text  \
0  When late night munchies strike, there's no better place to be. The fresh fish in the sushi is outstanding and they also have a large number of imports from Tsukiji. Read more on WinstonWanders below!   

   agreeCount  disagreeCount                        id user.firstName  \
0          12              0  5523f574498e1cb7c96aaebc           Mike   

  user.lastName  user.id  
0             W  4083549

The review gave glowing reviews to the food offered by this restaurant, highlighting its freshness. This comment has 12 upvotes, the highest among these restaurants. It seems that this restaurant offer quality food and if I am craving for some Japanese food, this is the place to go. 

# Results and Discussions

In [54]:
# Combining all five reviews (1 within 300m and 4 within 1km)
pd.concat([tips_filtered, tips_filtered0, tips_filtered1, tips_filtered2, tips_filtered3], ignore_index=True)

text  \
0            You can almost always walk right in with no wait. My favorite items are the eggplant salad with shrimp and pork, and the peppercorn duck breast. Everything I've tasted has been good, though.   
1                    Order the Chicken Pho, it's the best in the Bay area with a broth as delicious as it gets. If your up for a challenge add Pho Ga Long to your broth mixing chicken giblets to the mix.   
2                                           AMAZING VIETNAMESE. I had the pho and the shrimp salad. Super duper tasty. They do take out and you'll have to queue for seats. Always room for one at the bar.   
3            You can almost always walk right in with no wait. My favorite items are the eggplant salad with shrimp and pork, and the peppercorn duck breast. Everything I've tasted has been good, though.   
4  When late night munchies strike, there's no better place to be. The fresh fish in the sushi is outstanding and they also have a large number of imports from Tsukiji. Read more on WinstonWanders below!   

   agreeCount  disagreeCount                        id user.firstName  \
0           3              0  56eb15b4498e7f8447ee7b78           Jane   
1           3              0  545406bb498e43d4ed60d015    Staff Picks   
2           6              0  56466ebf498e5219f1fc7a9f        Harriet   
3           3              0  56eb15b4498e7f8447ee7b78           Jane   
4          12              0  5523f574498e1cb7c96aaebc           Mike   

  user.lastName    user.id  
0             R  168464378  
1           NaN   86922724  
2             R  111578617  
3             R  168464378  
4             W    4083549

In [23]:
dataframe_filtered.sort_values(by=['distance'])

name  \
21                         Civic Restaurant and Bar   
24                        Golden Kim Tar Restaurant   
20                              Restaurant Caliente   
12                            Quintero's Restaurant   
0                           Turtle Tower Restaurant   
22                  Thai Idea Vegetarian Restaurant   
29                         Teo's Restaurant and Bar   
14                           Cafe Splash Restaurant   
7          Restaurant & Hotel Upholstery Specialist   
1                                 Tú Lan Restaurant   
6                              7 Mission Restaurant   
28                                 Restaurant Depot   
27        The Restaurant at the End of the Universe   
5                          Pakwan Indian Restaurant   
25                            A La Turca Restaurant   
8                                Chutney Restaurant   
26                       Cityscape Bar & Restaurant   
17  Eden's Mediterranean Turkish & Greek Restaurant   
18                       Java Indonesian Restaurant   
9                                   Restaurant Anzu   
2                       Basil Thai Restaurant & Bar   
11                              Yemeni's Restaurant   
13                                  Fino Restaurant   
4                 Ryoko's Japanese Restaurant & Bar   
10                   Don Ramon's Mexican Restaurant   
16         The Restaurant Collection Under The Dome   
3              The Chieftain Irish Pub & Restaurant   
19                                Darbar Restaurant   
15                           MKT Restaurant and Bar   
23                               B Restaurant & Bar   

                       categories            address  \
21                           Food  272 McAllister St   
24             Chinese Restaurant      434 Larkin St   
20                           Food   86 McAllister St   
12             Mexican Restaurant        393 Eddy St   
0           Vietnamese Restaurant      645 Larkin St   
22  Vegetarian / Vegan Restaurant        710 Polk St   
29             Chinese Restaurant         112 7th St   
14          Vietnamese Restaurant       750 Ellis St   
7          Furniture / Home Store      120 Willow St   
1           Vietnamese Restaurant           8 6th St   
6              Chinese Restaurant         150 7th St   
28                           None                NaN   
27                     Strip Club    935 Ofarrell St   
5               Indian Restaurant    501 Ofarrell St   
25             Turkish Restaurant       869 Geary St   
8               Indian Restaurant       511 Jones St   
26        New American Restaurant    333 Ofarrell St   
17       Mediterranean Restaurant       552 Jones St   
18          Indonesian Restaurant    3015 Geary Blvd   
9             Japanese Restaurant       222 Mason St   
2                 Thai Restaurant     1175 Folsom St   
11      Middle Eastern Restaurant     1098 Sutter St   
13             Italian Restaurant        624 Post St   
4                Sushi Restaurant      619 Taylor St   
10             Mexican Restaurant        225 11th St   
16        New American Restaurant                NaN   
3                       Irish Pub         198 5th St   
19              Indian Restaurant       1412 Polk St   
15                     Steakhouse      757 Market St   
23                         Lounge      720 Howard St   

                         crossStreet        lat         lng  \
21                               NaN  37.780578 -122.416420   
24    btwn Turk St & Golden Gate Ave  37.781925 -122.417351   
20                               NaN  37.781052 -122.413371   
12  btwn Leavenworth St & Wagner Aly  37.783265 -122.413685   
0                  btwn Eddy & Ellis  37.783989 -122.417764   
22                        at Eddy St  37.783395 -122.419141   
29                        Mission St  37.779084 -122.410911   
14              Btwn Polk and Larkin  37.784405 -122.418565   
7                                NaN 

Seems like Asian food is really popular in San Francisco, as all the top option are Asian Food (2 Vietnamese, 1 Thai, 1 Japanese). Considering the amount of Asian Students in UC Davis, there is definitely a demand for those. 

From the reviews, two observations could be made. The first observation that can be made is that there is a growing trend of popularity for Asian Food in San Francisco (2 Vietnamese, 1 Thai, 1 Japanese, and 1 Chinese potentially), or at least for the Civic Centre neighbourhood, as they made up the top restaurant list. 

Secondly, the reviews from the top options are quite decent, which are not surprising considering they are the top options. Hence, they could all be recommended with consideration of how long you would like to travel. This list of restaurants could be recommended to the interested students. 

# Conclusion

The purpose of this project was to identify the most optimal restaurant given the limitation of distance, in order to aid stakeholders (prospective students) in narrowing down their options mostly when they search for food in between classes, where time could be limited and travelling afar infeasible. By looking at the ratings of the restaurants within 300 metres and 1000 metres respectively, the stakeholders' search cost could potentially be reduced, where they can gauge what are the options available within proximity and choose where they would like to frequent most.

Yet, akin to one of the most favorite response of a data scientists would reply to any question, that is 'it depends'. Indeed, the final decision on the most optimal restaurant will depend on the subjectivity of the stakeholders, just like many real life scenarios. 

In addition, this report is a very casual look into the 'best restaurant' in the area. Certainly, there are many more factors that can be taken into consideration when conducting this project. As a starter, I could definitely take a more scientific approach to define and fine-tune the metrics used to measure the 'best restaurant'. For instance, I could calculate the mean of ratings across San Francisco area to find out the average rating, instead of using an arbitrary 8.0 as the defining metric. The metric could also be broken into more criterias, such as quality of service, quality of food, cleaniness, etc. While that is challenging as FourSquare does not seem to provide such information, some text mining could be done for sentiment analysis. Lastly, if I could pull more reviews from the query, it will definitely provide a more balanced look at the restaurants.  